quickwit/quickwit-search/src/thread_pool.rs

```diff
@@ -18,22 +18,39 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::fmt;
+use std::sync::Arc;
 
 use once_cell::sync::OnceCell;
 use quickwit_common::metrics::GaugeGuard;
+use tantivy::Executor;
 use tracing::error;
 
+static SEARCH_THREAD_POOL: OnceCell<Arc<Executor>> = OnceCell::new();
+
+fn build_executor() -> Arc<Executor> {
+    let rayon_pool = rayon::ThreadPoolBuilder::new()
+        .thread_name(|thread_id| format!("quickwit-search-{thread_id}"))
+        .panic_handler(|_my_panic| {
+            error!("task running in the quickwit search pool panicked");
+        })
+        .build()
+        .expect("Failed to spawn the spawning pool");
+    Arc::new(Executor::ThreadPool(rayon_pool))
+}
+
 fn search_thread_pool() -> &'static rayon::ThreadPool {
-    static SEARCH_THREAD_POOL: OnceCell<rayon::ThreadPool> = OnceCell::new();
-    SEARCH_THREAD_POOL.get_or_init(|| {
-        rayon::ThreadPoolBuilder::new()
-            .thread_name(|thread_id| format!("quickwit-search-{thread_id}"))
-            .panic_handler(|_my_panic| {
-                error!("task running in the quickwit search pool panicked");
-            })
-            .build()
-            .expect("Failed to spawn the spawning pool")
-    })
+    let executor_ptr = Arc::as_ptr(SEARCH_THREAD_POOL.get_or_init(build_executor));
```

Comment:
This code is correct, but could break if we modified a different part of the code.

It is sometimes acceptable (especially because it is in the same file, and private stuff), but still, I suspect this is unnecessary here.

Doesn't returning an `Arc` in search_thread_pool work?

Reply:
```rust
pub async fn run_cpu_intensive<F, R>(cpu_heavy_task: F) -> Result<R, Panicked>
where
    F: FnOnce() -> R + Send + 'static,
    R: Send + 'static,
{
    let span = tracing::Span::current();
    search_executor().spawn_blocking(move || {
        let _guard = span.enter();
        let mut active_thread_guard =
            GaugeGuard::from_gauge(&crate::SEARCH_METRICS.active_search_threads_count);
        active_thread_guard.add(1i64);
        cpu_heavy_task()
    }).await.map_err(|_| Panicked)
}
```